### PR TITLE
Feature/1669  drag and drop

### PR DESCRIPTION
**Pull request by Skylsmoi**
https://github.com/tracim/tracim/issues/789
https://github.com/tracim/tracim/issues/1669

## Checkpoints

The goal of this section is to help code reviewer to do required checks on this pull request. Please don't edit this list.
If one or more of these checkpoints are out of scope, please write below why they are.

- [x] Code is clear enoug
- [x] If there are FIXME in the code, then there are associated issues and issue is referenced in the FIXME
- [x] If there are TODO, NOTE or HACK in code, date and developer initials are present
- [x] Automated tests have been written (covering feature or non regression), or *at least* an issue has been created for test implementation

### CODE REVIEW

**Comment by Skylsmoi** [comment]
Tests have not been written.
See https://github.com/tracim/tracim/issues/1905

File: frontend/src/component/Sidebar/WorkspaceListItem.jsx [hunk 1]
```diff
@@ -49,7 +56,12 @@ const WorkspaceListItem = props => {
             color: color(GLOBAL_primaryColor).light() ? '#333333' : '#fdfdfd'
           }}
         >
-          {props.label.substring(0, 2).toUpperCase()}
+          {isDropActive
+            ? isDropAllowed && isDropAllowedOnWorkspaceRoot
+              ? <i className='fa fa-arrow-circle-down' />
+              : <i className='fa fa-times-circle' />
+            : props.label.substring(0, 2).toUpperCase()
```
Comment (basilelegal) [inline frontend/src/component/Sidebar/WorkspaceListItem.jsx]: sould be a function

File: frontend/src/component/Workspace/Folder.jsx [hunk 2]
```diff
@@ -1,23 +1,48 @@
...
 
 require('./Folder.styl')
 
 class Folder extends React.Component {
+  calculateIcon = (folder, isDropActive, draggedItem = null) => {
+    const isHoveringSelf = draggedItem && draggedItem.contentId === folder.id
+
+    if (!isDropActive || isHoveringSelf) {
+      if (folder.isOpen) return 'fa-folder-open-o'
+      else return 'fa-folder-o'
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/component/Workspace/Folder.jsx]: fixed

File: frontend/src/component/Workspace/Folder.jsx [hunk 3]
```diff
@@ -1,23 +1,48 @@
...
+    }
+
+    const isHoveringSelfParent = draggedItem && draggedItem.parentId === folder.id
+    const isHoveringChildrenFolder = draggedItem.subFolderIdList && draggedItem.subFolderIdList.some(subFolderId => folder.id === subFolderId)
+
+    if (isHoveringSelfParent || isHoveringChildrenFolder) return 'fa-times-circle primaryColorFont'
+    else return 'fa-arrow-circle-down primaryColorFont'
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/component/Workspace/Folder.jsx]: fixed

File: frontend/src/component/Sidebar/WorkspaceListItem.jsx [hunk 4]
```diff
@@ -49,7 +56,12 @@ const WorkspaceListItem = props => {
             color: color(GLOBAL_primaryColor).light() ? '#333333' : '#fdfdfd'
           }}
         >
-          {props.label.substring(0, 2).toUpperCase()}
+          {isDropActive
+            ? isDropAllowed && isDropAllowedOnWorkspaceRoot
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/component/Sidebar/WorkspaceListItem.jsx]: fixed

File: frontend/src/component/Workspace/Folder.jsx [hunk 5]
```diff
@@ -168,7 +211,53 @@ class Folder extends React.Component {
...
 
-export default withRouter(Folder)
+const folderDndTarget = {
+  drop: props => ({
+    workspaceId: props.folderData.idWorkspace,
+    parentId: props.folderData.id
+  })
+}
+
+const folderDndTargetCollect = (connect, monitor) => ({
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/component/Workspace/Folder.jsx]: renamed

File: frontend/src/container/WorkspaceContent.jsx [hunk 6]
```diff
@@ -321,6 +324,37 @@ class WorkspaceContent extends React.Component {
...
+    const fetchMoveContent = await props.dispatch(putContentItemMove(source, destination))
+    switch (fetchMoveContent.status) {
+      case 200:
+        const {dropEffect, ...actionDestination} = destination
+        props.dispatch(moveWorkspaceContent(source, actionDestination))
+        this.loadContentList(state.idWorkspaceInUrl)
+        break
+      case 400:
+        switch (fetchMoveContent.json.code) {
+          case 3002: props.dispatch(newFlashMessage(props.t('A content with same name already exists'), 'warning')); break
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/WorkspaceContent.jsx]: I partially agree with that
I did the change anyway

File: frontend/src/container/WorkspaceContent.jsx [hunk 7]
```diff
@@ -321,6 +324,37 @@ class WorkspaceContent extends React.Component {
     props.history.push(`${PAGE.WORKSPACE.NEW(state.idWorkspaceInUrl, contentType)}?${qs.stringify(newUrlSearch, {encode: false})}&parent_id=${idFolder}`)
   }
 
+  handleDropMoveContent = async (source, destination) => {
+    const { props, state } = this
+
+    if (source.workspaceId === destination.workspaceId && source.parentId === destination.parentId) return
+    if (source.contentId === destination.parentId) return
+    if (source.subFolderIdList && source.subFolderIdList.some(subFolderId => destination.parentId === subFolderId)) return
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/container/WorkspaceContent.jsx]: it is.
I have tested it

File: frontend/src/component/Sidebar/WorkspaceListItem.jsx [hunk 8]
```diff
@@ -5,102 +5,138 @@ import classnames from 'classnames'
...
 import PropTypes from 'prop-types'
 import AnimateHeight from 'react-animate-height'
+import { DropTarget } from 'react-dnd'
+import {DRAG_AND_DROP, ROLE_OBJECT} from '../../helper.js'
 
 const qs = require('query-string')
 
-const shouldDisplayAsActive = (location, idWorkspace, activeIdWorkspace, app) => {
-  if (idWorkspace !== activeIdWorkspace) return false
+class WorkspaceListItem extends React.Component {
```
Comment (Skylsmoi) [inline frontend/src/component/Sidebar/WorkspaceListItem.jsx]: I have changed this stateless component to a statefull component.
That's why there is a big diff on it

File: frontend/src/helper.js [hunk 9]
```diff
@@ -83,13 +83,12 @@ export const ROLE = [{
   label: 'Reader' // label must be used in components
 }]
 
-export const findIdRoleUserWorkspace = (idUser, memberList, roleList) => {
+export const findUserWorkspaceRoleId = (idUser, memberList, roleList) => {
   const myself = memberList.find(u => u.id === idUser) || {role: 'reader'}
```
[the remark this answers is not in the document]
Comment (Skylsmoi) [reply frontend/src/helper.js]: fixed

File: frontend/src/component/Sidebar/WorkspaceListItem.jsx (hunk 1 above)
Comment (Skylsmoi) [reply frontend/src/component/Sidebar/WorkspaceListItem.jsx]: It is the only place where we need this
I have moved this ternary into a specific function `getIcon()` to be clearer

**Pull request merged**